### ISSUE TITLE
`attrs` not working for json-api-serializer

### ISSUE THREAD

**Issue by digabit** [emberjs/data#3810]
Defining an `attrs` map when using JSONAPISerializer does not seem to be working.

For example if I have the following model:

```javascript
// app/models/product.js
export default DS.Model.extend({
	productNumber: DS.attr('string')
});
```

and the following serializer:

```javascript
// app/serializers/product.js
export default DS.JSONAPISerializer.extend({
  attrs: {
    productNumber: 'partNumber'
  }
});
```

and JSON:API data like this coming over the wire:

```JSON
{
	"data": [{
		   "type": "products",
		   "id": "15296",
		   "attributes": {
			   "partNumber": "3004-787"
		   },
		   "relationships": ...,
		   "links": ...
        }]
}
```

The serializer does not assign '3004-787' to the `productNumber` property of the new `product`.

The root cause seems to be in `json-api-serializer` which uses `normalizeUsingDeclaredMapping` from `json-serializer`. That method looks for the keys on `attrs` at the root level of the `hash`. In JSON:API these properties are inside the `attributes` property. So it seems like we need to override `normalizeUsingDeclaredMapping` and do something like this:

```javascript
// serializers/product.js
 normalizeUsingDeclaredMapping: function (typeClass, hash) {
    var get = Ember.get;
    var attrs = get(this, "attrs");
    var payloadKey, key;

    if (attrs) {
      for (key in attrs) {
        payloadKey = this._getMappedKey(key);
        if (!hash.attributes || !hash.attributes.hasOwnProperty(payloadKey)) {
          continue;
        }

        if (payloadKey !== key) {
          hash.attributes[key] = hash.attributes[payloadKey];
          delete hash.attributes[payloadKey];
        }
      }
    }
  }
```

This does in fact solve the problem in my environment. I didn't submit this as a pull request because I think there are probably other areas which I'm not familliar with but which need to be fixed as well. Like when serializing this property we need to do the reverse to construct the proper JSON from the model. Also does `attrs` apply to relationship names as well? Seems like it might have the same problem. Thanks

**Comment by HeroicEric** [comment]
@digabit If your server is sending the key as `partNumber`, I think you'll want to use:

```javascript
export default DS.JSONAPISerializer.extend({
  attrs: {
    productNumber: { key: 'partNumber' }
  }
});
```

**Comment by HeroicEric** [comment]
Or maybe you need to do something like:

```javascript
import DS from 'ember-data';

const { JSONAPISerializer } = DS;

export default JSONAPISerializer.extend({
  keyForAttribute(attr) {
    if (attr === 'partNumber') {
      attr = 'productNumber';
    }

    return this._super(...arguments);
  }
});
```

**Comment by digabit** [comment]
Thanks for the suggestions. Your first idea doesn't seem to work. Your second suggestion does give another way to do it although you need to:

 ```javascript
return this._super(attr);
```

However what you are proposing are work-arounds. This is exactly what the `attrs` mapping is supposed to for and it doesn't seem to be working.

**Comment by pangratz** [comment]
@digabit I think you should open a PR so this is fixed in the `JSONAPISerializer`. The `normalizeUsingDeclaredMapping` normalizes the keys for attributes and relationships, so this needs to be considered in the fix. I would change [this line](https://github.com/emberjs/data/blob/4745dcb6fa00664ddb19bab9f4913695656f7e1c/packages/ember-data/lib/serializers/json-api-serializer.js#L279) with something like:

```
this.normalizeUsingDeclaredMapping(modelClass, resourceHash.attributes);
this.normalizeUsingDeclaredMapping(modelClass, resourceHash.relationships);
```

There should likely be checks added, since `normalizeUsingDeclaredMapping` expects a hash as second argument, but `attributes` or `relationships` might be undefined on the JSONAPI document.

**Comment by digabit** [comment]
@pangraz Ah yeah that is cleaner than overriding the whole method. Alright I'll put together a PR.

**Comment by HeroicEric** [comment]
@digabit my bad. I misunderstood. I didn't realize `normalizeUsingDeclaredMapping` was a thing. I thought you added that.

**Comment by digabit** [comment]
@HeroicEric No problem!

**Comment by pangratz** [comment]
Hey @digabit, kindly asking if you had any chance yet for the PR?

**Comment by weinberg** [comment]
Hi @pangratz - just filed the PR here: https://github.com/emberjs/data/pull/3845. Thanks

**Comment by bmac** [comment]
I believe this issue is solved by #3847.
